// @paperclipai/shared/dist/gtm.js
function humanizeToken(value) {
    return value
        .replace(/[_-]+/g, " ")
        .replace(/\s+/g, " ")
        .trim()
        .replace(/\b\w/g, (match) => match.toUpperCase());
}
export function mapGtmKnowledgeKind(sourceType) {
    return sourceType === "table" ? "table" : "item";
}
export function formatConnectorLabel(type) {
    if (!type?.trim())
        return "Manual";
    return humanizeToken(type);
}
export function formatKnowledgeGroupLabel(table) {
    return humanizeToken(table.fileName);
}
export function createDefaultGtmState(now = new Date().toISOString()) {
    return {
        profile: {
            growthubAccountEmail: null,
            workspaceName: "Workspace",
            ghAppPath: null,
            workspaceId: null,
            adminId: null,
        },
        knowledge: {
            table: {
                id: "workspace-knowledge-table",
                agentSlug: "__user_general",
                compressed: false,
                createdAt: now,
                fileName: "workspace_knowledge",
                isActive: true,
                itemCount: 0,
                metadata: {
                    origin: "table",
                    notes: "No workspace knowledge has been imported yet.",
                },
                sourceType: "table",
                storagePath: "knowledge/tables/workspace_knowledge.txt",
                updatedAt: now,
                userId: "local-user",
            },
            items: [],
        },
        connectors: [],
        workflow: {
            id: "local-sdr",
            label: "Local SDR",
            runnerPath: null,
            referenceInterfaces: {
                internalSocialsPath: null,
                localSdrPath: null,
            },
            lastRun: {
                command: null,
                error: null,
                finishedAt: null,
                pid: null,
                startedAt: null,
                status: "idle",
            },
        },
    };
}
export function coerceGtmState(raw) {
    const fallback = createDefaultGtmState();
    if (!raw || typeof raw !== "object")
        return fallback;
    const candidate = raw;
    const fallbackItemTemplate = fallback.knowledge.table;
    const fallbackConnectorTemplate = fallback.connectors[0] ?? {
        id: "gtm-connector",
        label: "Connector",
        target: "Configure workspace and connector settings",
        status: "needs_attention",
        config: {
            workspaceId: null,
            adminId: null,
        },
    };
    return {
        profile: {
            ...fallback.profile,
            ...candidate.profile,
        },
        knowledge: {
            table: {
                ...fallback.knowledge.table,
                ...candidate.knowledge?.table,
                metadata: {
                    ...fallback.knowledge.table.metadata,
                    ...candidate.knowledge?.table?.metadata,
                },
            },
            items: Array.isArray(candidate.knowledge?.items) && candidate.knowledge.items.length > 0
                ? candidate.knowledge.items.map((item, index) => ({
                    ...fallbackItemTemplate,
                    ...item,
                    metadata: {
                        ...fallbackItemTemplate.metadata,
                        ...item.metadata,
                    },
                }))
                : fallback.knowledge.items,
        },
        connectors: Array.isArray(candidate.connectors) && candidate.connectors.length > 0
            ? candidate.connectors.map((connector, index) => ({
                ...fallbackConnectorTemplate,
                ...connector,
                config: {
                    ...fallbackConnectorTemplate.config,
                    ...connector.config,
                },
            }))
            : fallback.connectors,
        workflow: {
            ...fallback.workflow,
            ...candidate.workflow,
            referenceInterfaces: {
                ...fallback.workflow.referenceInterfaces,
                ...candidate.workflow?.referenceInterfaces,
            },
            lastRun: {
                ...fallback.workflow.lastRun,
                ...candidate.workflow?.lastRun,
            },
        },
    };
}
export function toGtmViewModel(state) {
    const workflowRun = state.workflow.lastRun;
    return {
        profile: {
            account: state.profile.growthubAccountEmail ?? "Not connected",
            workspace: state.profile.workspaceName ?? "Workspace",
            appConnection: state.profile.ghAppPath ?? "App not configured",
        },
        knowledge: {
            group: {
                label: formatKnowledgeGroupLabel(state.knowledge.table),
                connector: formatConnectorLabel(state.knowledge.table.metadata.connector_type),
                notes: state.knowledge.table.metadata.notes ?? null,
                itemCount: state.knowledge.items.length,
            },
            items: state.knowledge.items.map((item) => ({
                title: item.fileName,
                kind: mapGtmKnowledgeKind(item.sourceType),
                notes: item.metadata.notes ?? null,
            })),
        },
        connectors: state.connectors.map((connector) => ({
            label: connector.label,
            target: connector.target,
            status: connector.status === "connected" ? "Connected" : "Needs attention",
            summary: connector.config.workspaceId
                ? `Workspace ${connector.config.workspaceId}`
                : "Connector requires configuration",
        })),
        workflow: {
            label: state.workflow.label,
            runner: state.workflow.runnerPath ?? "No local runner configured",
            status: !state.workflow.runnerPath
                ? "Needs setup"
                : workflowRun.status === "running"
                    ? "Running"
                    : workflowRun.status === "failed"
                        ? "Needs attention"
                        : "Ready",
            lastRunAt: workflowRun.startedAt ?? "Not started",
            actionLabel: state.workflow.runnerPath ? "Launch Local SDR" : "Configure Local SDR",
            interfaces: [
                state.workflow.referenceInterfaces.internalSocialsPath ?? "Reference app not configured",
                state.workflow.referenceInterfaces.localSdrPath ?? "Workflow runner not configured",
            ],
        },
    };
}
