// @paperclipai/shared/dist/surface-runtime.js
let runtimeContract = null;
function buildSurfaceRuntimeContract(profile = "dx") {
    if (profile === "gtm") {
        return Object.freeze({
            profile: "gtm",
            mounts: Object.freeze({
                routes: "/gtm",
                cli: "gtm",
            }),
            capabilities: Object.freeze({
                dxEnabled: false,
                gtmEnabled: true,
            }),
        });
    }
    return Object.freeze({
        profile: "dx",
        mounts: Object.freeze({
            routes: "/dx",
            cli: "dx",
        }),
        capabilities: Object.freeze({
            dxEnabled: true,
            gtmEnabled: false,
        }),
    });
}
export function initializeSurfaceRuntimeContract(profile = "dx") {
    if (runtimeContract && runtimeContract.profile === profile)
        return runtimeContract;
    runtimeContract = buildSurfaceRuntimeContract(profile);
    return runtimeContract;
}
export function getSurfaceRuntimeContract() {
    if (!runtimeContract) {
        throw new Error("Surface runtime contract accessed before bootstrap initialization");
    }
    return runtimeContract;
}
